IOS/Network: Emulate socket fd table

sepalani opened the pull request:
This PR tries to emulate a Wii fd table. Dolphin was using the host's fd when emulating sockets, while it shouldn't be a problem, it seems that in some cases (like the Netflix channel) there is a check against that fd value.

Regarding the Netflix channel, if the fd value is greater than 23 it returns a EBADF error (which consequently is sent to SO commands like CONNECT). I assume, it's probably using select/poll to monitor sockets and that it does a sanity check against a specified limit (```max_fd```?).

~Currently, the fd values are computed empirically. It will need a hardware test to confirm that it matches the Wii's behaviour. It also assumes that it has a separate table for other (non-socket) fd generated. Otherwise, the fd table will need to be implemented somewhere else in IOS code.~ Not the case anymore, thanks to @shuffle2.

## TODO
Based on @shuffle2's gist: https://gist.github.com/shuffle2/4541e4e96ba3e04a34bd36b0e07d3013#file-so_socket-c-L71
- [x] Create the socket fd table
- [x] Replace logs using host socket fd with wii fd
- [x] Handle SO_ENETRESET (39)
- [x] Handle SO_EAFNOSUPPORT (5)
- [x] Handle SO_EPROTONOSUPPORT (68)
- [x] Handle SO_EMFILE (33)
- [x] Handle SO_EPROTOTYPE (69)
- [x] Handle SO_ENOMEM (49)

Ready to be reviewed & merged.

leoetlino commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -604,23 +605,35 @@ void WiiSocket::DoSock(Request request, SSL_IOCTL type)
   if (fd >= 0)
   {
-    WiiSocket& sock = WiiSockets[fd];
+    s32 wii_fd = 3;
```
Is there a limit in IOS?

sepalani replied:
No idea. That's why it needs a hardware test to know at least the following:
 - the max_fd value (if any)
 - the maximum number of open sockets the Wii can support
 - how are fd value given and does it have a separate table for sockets

shuffle2 replied:
You should just be able to look at IOS disasm to figure this out, which would give a better understanding anyways.

BhaaLseN commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -231,6 +231,7 @@ void WiiSocket::Update(bool read, bool write, bool except)
         WiiSockMan::Convert(*wii_name, local_name);
 
         int ret = connect(fd, (sockaddr*)&local_name, sizeof(local_name));
+        ret = WiiSockMan::GetInstance().AddSocket(ret);
```
`connect` return 0 on success or -1 on error; it doesn't return a socket. Did you want to call this on `fd` instead of `ret`?

sepalani replied:
Good catch, that's an oversight. It shouldn't be called at all.

lioncash commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -604,23 +605,35 @@ void WiiSocket::DoSock(Request request, SSL_IOCTL type)
   return ret;
 }
 
+s32 WiiSockMan::GetHostSocket(s32 wii_fd)
```
this can be a const member function using `return WiiSockets.at(wii_fd).fd` in your if statement body.

riking commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -604,23 +604,35 @@ void WiiSocket::DoSock(Request request, SSL_IOCTL type)
+{
+  if (WiiSockets.count(wii_fd) > 0)
+    return WiiSockets.at(wii_fd).fd;
+  return EBADF;
```
This should be `return -EBADF;`, no?

```
$ cat > test.c
#include <errno.h>
#include <stdio.h>
int main() { printf("%d\n", EBADF); }
$ gcc test.c ; ./a.out
9
```

Also, this code is missing a lot of `errno` checks.

sepalani replied:
I borrowed this value from the ```DeleteSocket()``` member function, so I'm not sure. You might be right.

riking commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -239,25 +239,25 @@ void WiiSocket::Update(bool read, bool write, bool except)
         }
 
-        WiiSockMan::GetInstance().AddSocket(ReturnValue);
+        ret = WiiSockMan::GetInstance().AddSocket(ret);
```
The return value should be checked for -1/error before being added to the socket table.

sepalani replied:
It's checked in the ```AddSocket() method``` and is the behaviour Dolphin had before this PR.

riking commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -604,23 +604,35 @@ void WiiSocket::DoSock(Request request, SSL_IOCTL type)
-  s32 ret = GetNetErrorCode(fd, "NewSocket", false);
-  AddSocket(ret);
+  s32 fd = static_cast<s32>(socket(af, type, protocol));
+  s32 wii_fd = AddSocket(fd);
```
again, check for errors from the host `socket()` call and forward to the wii errno

sepalani replied:
Ditto.

shuffle2 commented:
Here is the related SO code from IOS56 which is used to lookup the object via `fd`:
```c
  v15 = irq_disable();
  entry = 0;
  if ( fd <= 23 )
  {
    entry = &sock_fd_table[fd];
    if ( entry->refcount > 0 && (obj = entry->obj) != 0 )
    {
      ++entry->refcount;
      if ( pobj )
        *pobj = obj;
    }
    else
    {
      entry = 0;
    }
  }
  irq_restore(v15);
  return entry;
```
So the hardcoded size check seems correct.

shuffle2 commented:
Here is how the fd is selected by IOCTL_SO_SOCKET: https://gist.github.com/shuffle2/4541e4e96ba3e04a34bd36b0e07d3013#file-so_socket-c-L101
basically, SO keeps it's own fd table, and it returns the first currently free (refcount==0) index into the table as the fd.

shuffle2 commented:
I updated the above gist to clear up confusion around the "protocol" arg for SOSocket()...the params follow standard posix socket() https://gist.github.com/shuffle2/4541e4e96ba3e04a34bd36b0e07d3013#file-so_socket-c-L71

sepalani commented:
@riking you were correct concerning the error values. It was a mistake from the original code because ```DeleteSocket``` returns the Wii error code via the ```CloseFd()``` member function. So I addressed your comment.

I tested this PR with games using NintendoWFC (with altwfc) and other channels and they don't hit the fd maximum limit. That is to say, they apparently don't forget to close sockets.

Otherwise, should I change the constants used with their respective define value such as AF_INET and so on or leave a comment about them?

Regardless, this PR is ready to be reviewed & merged.

shuffle2 commented:
> Otherwise, should I change the constants used with their respective define value such as AF_INET and so on or leave a comment about them?

What do you mean by that?

sepalani commented:
@shuffle2 
Replace this:
```C++
if (type != 1 && type != 2)
```
With this or something along those lines:
```C++
if (type != 1 && type != 2)  // SOCK_STREAM && SOCK_DGRAM
```

shuffle2 commented:
A comment would be fine - I'm not sure how much we have to care about those values differing on host platforms, best to be on the safe side I guess.
